internal/archive/extractor_test.go

```diff
@@ -0,0 +1,224 @@
+package archive
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"runtime"
+	"testing"
+
+	"github.com/romie-gr/romie/internal/utils"
+)
+
+var (
+	existingFolder = "./testdata"
+	extractionPath = "./testdata/archive"
+	extractToPath  = "./testdata/new-destination"
+	nonWritableDir = "/sys"
+
+	missingArchive = "./testdata/missing.zip"
+	nonArchiveFile = "./testdata/archive.txt"
+
+	invalidZip = "./testdata/invalid.zip"
+	validZip   = "./testdata/archive.zip"
+
+	invalid7z = "./testdata/invalid.7z"
+	valid7z   = "./testdata/archive.7z"
+
+	invalidRar = "./testdata/invalid.rar"
+	validRar   = "./testdata/archive.rar"
+)
+
+func ExampleExtract() {
+	err := Extract("/path/to/archive.zip")
+	if err != nil {
+		fmt.Printf("Zip archive extraction failure: %v", err)
+	} else {
+		fmt.Println("Zip archive was successfully extracted")
+	}
+	// Output: Zip archive extraction failure: file /path/to/archive.zip not found
+}
+
+func TestExtract(t *testing.T) {
+	tests := []struct {
+		name    string
+		source  string
+		wantErr bool
+	}{
+		{
+			"Returns error if file does not exist",
+			missingArchive,
+			true,
+		},
+		{
+			"Returns error if file is a directory",
+			existingFolder,
+			true,
+		},
+		{
+			"Returns error if file not a valid archive",
+			nonArchiveFile,
+			true,
+		},
+		{
+			"Zip: Returns error if file not a valid archive",
+			invalidZip,
+			true,
+		},
+		{
+			"7z: Returns error if file not a valid archive",
+			invalid7z,
+			true,
+		},
+		{
+			"Rar: Returns error if file not a valid archive",
+			invalidRar,
+			true,
+		},
+		{
+			"Zip: Returns no error if file is a valid archive",
+			validZip,
+			false,
+		},
+		{
+			"7z: Returns no error if file is a valid archive",
+			valid7z,
+			false,
+		},
+		{
+			"Rar: Returns no error if file is a valid archive",
+			validRar,
+			false,
+		},
+	}
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			if err := Extract(tt.source); (err != nil) != tt.wantErr {
+				t.Errorf("Extract(%q) error = %v, wantErr %v", tt.source, err, tt.wantErr)
+			}
+
+			// Assert file contents and cleanup
+			if !tt.wantErr {
+				followUpAssertAndCleanUp(t, extractionPath)
+			}
+		})
+	}
+}
+
+func ExampleExtractTo() {
+	err := ExtractTo("/path/to/archive.zip", "/destination/folder")
+	if err != nil {
+		fmt.Printf("Zip archive extraction failure: %v", err)
+	} else {
+		fmt.Println("Zip archive was successfully extracted")
+	}
+	// Output: Zip archive extraction failure: file /path/to/archive.zip not found
+}
+
+func TestExtractTo(t *testing.T) {
+	type args struct {
+		source      string
+		destination string
+	}
+
```

Comment:
If this running on a Windows machine, replace the the `nonWritableDir` with `C:/Windows`. In that folder you need Administrator access to write stuff.

```suggestion

	if runtime.GOOS == "windows" {
		nonWritableDir = windowsDir(t)
	}
	
```